### ISSUE TITLE
heap crash

### ISSUE THREAD

**Issue by OnlyOneD** [Sandrem/FlyCasual#549]
Mostly while editing squads or making new ones...it seems to tends to try and load everything and going above 1.6GB of memory and crashes.

Noticed it mostly happening with AI vs AI .. even while not editing squads it sometimes happens.

**Comment by Sandrem** [comment]
Unfortunately, I still cannot reproduce this behaviour on my PC.
Please, if you will have information what steps cause crash or what actions cause massive memory usage - inform me. I cannot fix this problem without such information.

Also, following information can be useful: OS (version, x86 or x64), amount of RAM installed, antivirus

**Comment by OnlyOneD** [comment]
Using Win10 64bit on a I7 with 24GB RAM and a nVidia gtx970. (Pagefile system managed, currently at 4GB)

(Teams in XWS for ease below)
Starting - Beautiful quality - AI vs AI
5* A-wing - Green Wing - Test Pilot - Chardaan refit - Predator
vs
5* TIE Interceptor - Avenger Squad

-player 1 wins, restart option: crash while reloading!

-restarting game, fast quality (around 303MB-605 ingame), load same teams:
player 1 wins, 
restart option: stuck at loading screen while loading from 600MB to 1.651MB, then crash!


Seems every time a ships fires it allocates a bit of memory, but some gets released after all shooting is resolved in a round. Moving however also allocates memory, but it doesn't seem to get released and as turns go by, more and more memory is allocated/used (even though it's the same ships and nothing really new gets into the game, other than some damage, Target locks etc.)

With smaller squads (say 2 vs 3) it starts at memory usage of previous match (from what it was at the end, not when the match first started) and doesn't seem to keep loading while using under 600-650MB for some reason. After a while of restarting with the same team, it will still crash in the end though. After 3rd game it hits 650MB used memory..and kept loading again until 1600MB and crashes.

In another game with 5vs2 it kept going for quite a while, after I went back to menu around 600MB (after 2nd game/1 restart option), even at 900MB (4th game since starting) it still ran. 5th game (3rd restart option chosen since going back to menu) it crashed again while reloading though (1200MB at that time)

A-Wings XWS:

{"name":"5 a-wing","faction":"rebel","points":100,"version":"0.3.0","pilots":[{"name":"greensquadronpilot","points":20,"ship":"awing","upgrades":{"title":["awingtestpilot"],"missile":["chardaanrefit"],"ept":["predator"]},"vendor":{"Sandrem.FlyCasual":{"skin":"Green"}}},{"name":"greensquadronpilot","points":20,"ship":"awing","upgrades":{"title":["awingtestpilot"],"missile":["chardaanrefit"],"ept":["predator"]},"vendor":{"Sandrem.FlyCasual":{"skin":"Green"}}},{"name":"greensquadronpilot","points":20,"ship":"awing","upgrades":{"title":["awingtestpilot"],"missile":["chardaanrefit"],"ept":["predator"]},"vendor":{"Sandrem.FlyCasual":{"skin":"Green"}}},{"name":"greensquadronpilot","points":20,"ship":"awing","upgrades":{"title":["awingtestpilot"],"missile":["chardaanrefit"],"ept":["predator"]},"vendor":{"Sandrem.FlyCasual":{"skin":"Green"}}},{"name":"greensquadronpilot","points":20,"ship":"awing","upgrades":{"title":["awingtestpilot"],"missile":["chardaanrefit"],"ept":["predator"]},"vendor":{"Sandrem.FlyCasual":{"skin":"Green"}}}],"description":"Green Squadron Pilot + A-Wing Test Pilot + Chardaan Refit + Predator
Green Squadron Pilot + A-Wing Test Pilot + Chardaan Refit + Predator
Green Squadron Pilot + A-Wing Test Pilot + Chardaan Refit + Predator
Green Squadron Pilot + A-Wing Test Pilot + Chardaan Refit + Predator
Green Squadron Pilot + A-Wing Test Pilot + Chardaan Refit + Predator"}


TIE Interceptors XWS:

{"name":"5 Interceptors","faction":"imperial","points":100,"version":"0.3.0","pilots":[{"name":"avengersquadronpilot","points":20,"ship":"tieinterceptor","upgrades":{},"vendor":{"Sandrem.FlyCasual":{"skin":"Blue"}}},{"name":"avengersquadronpilot","points":20,"ship":"tieinterceptor","upgrades":{},"vendor":{"Sandrem.FlyCasual":{"skin":"Blue"}}},{"name":"avengersquadronpilot","points":20,"ship":"tieinterceptor","upgrades":{},"vendor":{"Sandrem.FlyCasual":{"skin":"Blue"}}},{"name":"avengersquadronpilot","points":20,"ship":"tieinterceptor","upgrades":{},"vendor":{"Sandrem.FlyCasual":{"skin":"Blue"}}},{"name":"avengersquadronpilot","points":20,"ship":"tieinterceptor","upgrades":{},"vendor":{"Sandrem.FlyCasual":{"skin":"Blue"}}}],"description":"Avenger Squadron Pilot
Avenger Squadron Pilot
Avenger Squadron Pilot
Avenger Squadron Pilot
Avenger Squadron Pilot"}

**Comment by Sandrem** [comment]
Thank you very much!

**Comment by OnlyOneD** [comment]
I've googled a bit with "unity engine new heap overflow" because I like using AI vs AI to try out builds .. even though most large ships tend to get off the map in 2-3 turns. And the following seems to explain a little about the problem: https://forum.unity.com/threads/fatal-error-in-gc-too-many-heap-sections.58746/

And following link seems to explain how resolve it, it seems that.. unless a maximum buffer is stated with new it will just keep on buffing and not recycling old/abandoned stuff like it should. Quoted from website below:

//
1 buffer = new ArrayBuffer(TOTAL_MEMORY)

TOTAL_MEMORY is defined by WebGL Memory Size in the Player Settings. The default value is 256mb, but that’s just an arbitrary value we chose. In fact an empty project works with just 16mb.

However, real-world content will likely need more, something like 256 or 386mb in most cases. Keep in mind that the more memory is needed, the fewer end-users will be able to run it.
// from:
https://blogs.unity3d.com/2016/09/20/understanding-memory-in-unity-webgl/


Personally I think 512MB would be a decent cap for now, only seen 'normal' memory usage go above 400MB when watching a lot of ships/cards etc in the squad builder (went up to 850MB orso and it wouldn't need all that data all the time), unless when that error heap overflow happened like mentioned in 1st post.

Hope it helpes, sadly not really grounded in C++ myself, wish I could help more.
